README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# squire

<!-- badges: start -->
[![Project Status: WIP – Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
[![R build status](https://github.com/mrc-ide/squire/workflows/R-CMD-check/badge.svg)](https://github.com/mrc-ide/squire/actions)
[![Codecov test coverage](https://codecov.io/gh/mrc-ide/squire/branch/master/graph/badge.svg)](https://codecov.io/gh/mrc-ide/squire?branch=master)
<!-- badges: end -->

squire enables users to simulate models of SARS-CoV-2 epidemics. This is
done using an age-structured SEIR model that also explicitly considers
healthcare capacity and disease severity.

## IMPORTANT NOTES

:warning: This code is released with no support. Please submit any
questions or bugs as [issues](https://github.com/mrc-ide/squire/issues)
and we will try to address them as quickly as possible.

:warning: This model is in active development and so parameter name and
behaviours, and output file formats will change without notice.

:warning: The model is stochastic. Multiple runs with different seeds
should be undertaken to see average behaviour.

:warning: As with any mathematical model, it is easy to misconfigure
inputs and therefore get meaningless outputs. Please contact the authors
if you intend to publish results using `squire`.

## Overview

squire is a package enabling users to quickly and easily generate
calibrated estimates of SARS-CoV-2 epidemic trajectories under different
control scenarios. It consists of the following:

  - An age-structured SEIR model incorporating explicit passage through
    healthcare settings and explicit progression through disease
    severity stages.
  - The ability to calibrate the model to different epidemic start-dates
    based on available death data.
  - Simulate the impacts of different control interventions (including
    general social distancing, specific shielding of elderly
    populations, and more stringent suppression strategies).

If you are new to squire, the best place to start is below, where we
detail how to install the package, how to set up the model, and how to
run it with and without control
interventions.

## Model Structure

### Overall Structure

<img src="https://raw.githubusercontent.com/mrc-ide/squire/master/images/Explicit_Healthcare_Model_Structure.JPG" align="center" style = "border: none; float: center;" width = "600px">

squire uses an age-structured SEIR model, with the infectious class
divided into different stages reflecting progression through different
disease severity pathways. These compartments are:  
\* S = Susceptibles  
\* E = Exposed (Latent Infection)  
\* I<sub>Mild</sub> = Mild Infections (Not Requiring Hospitalisation)  
\* I<sub>Case</sub> = Infections Requiring Hospitalisation  
\* I<sub>Hospital</sub> = Hospitalised (Requires Hospital Bed)  
\* I<sub>ICU</sub> = ICU (Requires ICU Bed)  
\* I<sub>Rec</sub> = Recovering from ICU Stay (Requires Hospital Bed)  
\* R = Recovered  
\* D =
Dead

### Decision Trees for Healthcare Capacity

<img src="https://raw.githubusercontent.com/mrc-ide/squire/master/images/Explicit_Healthcare_Oxygen_Decision_Tree.JPG" align="center" style = "border: none; float: center;" width = "400px">

Given initial inputs of hospital/ICU bed capacity and the average time
cases spend in hospital, the model dynamically tracks available hospital
and ICU beds over time.

Individuals newly requiring hospitalisation (either a hospital or ICU
bed) are then assigned to either receive care (if the relevant bed is
available) or not (if maximum capacity would be exceeded otherwise).
Whether or not an individual receives the required care modifies their
probability of dying.

## Installation

<i>squire</i> utilises the package
[odin](https://github.com/mrc-ide/odin) to generate the model.
[odin](https://github.com/mrc-ide/odin) implements a high-level language
for implementing mathematical models and can be installed by running the
following command:

``` r
install.packages("odin")
```

The model generated using odin is written in C and so you will require a
compiler to install dependencies for the package and to build any models
with odin. Windows users should install
[Rtools](https://cran.r-project.org/bin/windows/Rtools/). See the
relevant section in
[R-admin](https://cran.r-project.org/doc/manuals/r-release/R-admin.html#The-Windows-toolset)
for advice. Be sure to select the “edit PATH” checkbox during
installation or the tools will not be found.

The function `odin::can_compile()` will check if it is able to compile
things, but by the time you install the package that will probably have
been satisfied.

After installation of odin, ensure you have the devtools package
installed by running the following:

``` r
install.packages("devtools")
```

Then install the <i>squire</i> package directly from GitHub by running:

``` r
devtools::install_github("mrc-ide/squire")
```

If you have any problems installing then please raise an issue on the
<i>squire</i> [`GitHub`](https://github.com/mrc-ide/squire/issues).

If everything has installed correctly, we then need to load the
package:

``` r
library(squire)
```

## Getting Started

### Running the Model (Unmitigated)

### 1\. Running the model using baseline parameters and no control interventions

The full model is referred to as the **explicit\_SEEIR** model, with
hospital pathways explicitly exploring whether individuals will require
a general hospital bed providing oxygen or an ICU bed that provides
ventilation.

To run the model we need to provide at least one of the following
arguments:

  - `country`
  - `population` and `contact_matrix_set`

If the `country` is provided, the `population` and `contact_matrix_set`
will be generated (if not also specified) using the demographics and
matrices specified in the [global
report](https://www.imperial.ac.uk/mrc-global-infectious-disease-analysis/covid-19/report-12-global-impact-covid-19/).

To run the model by providing the `country` we use
`run_explicit_SEEIR_model()`:

``` r
r <- run_explicit_SEEIR_model(country = "Afghanistan")
```

The returned object is a `squire_simulation` object, which is a list of
two ojects:

  - `output` - model output
  - `parameters` - model parameters

`squire_simulation` objects can be plotted as follows:

``` r
plot(r)
```

<img src="man/figures/README-base plot-1.png" width="100%" />

This plot will plot each of the compartments of the model output. We can
also plot specific compartments using the `var_select` argument that can
be passed to `plot()`. Arguments passed to `var_select` must be one of
the variables in the plot
above.

``` r
plot(r, var_select = c("E", "IMild"))
```

<img src="man/figures/README-subset variables plot-1.png" width="100%" />
Or, you can specify one of `deaths`, `infections`, `hospital_occupancy`,
`ICU_occupancy`, `hospital_demand` or `ICU_demand`, and plot these
summary metrics that represent the combination of a number of different
compartment
e.g:

``` r
plot(r, var_select = "deaths")
```

<img src="man/figures/README-subset variables plot2-1.png" width="100%" />

All of the plotting above makes uses of the `squire` function
`format_output` which provides you with a means of manipulating and
managing the output from a `run_explicit_SEEIR_model` call. Using it you
can specify the model outputs (e.g. compartments) you want, as well as
whether you want that output aggregated over age or not. Here we extract
the latent compartment (E). The data columns correspond to the
compartment name (`compartment`), timestep (`t`), model run number
(`replicate`) and the model output (`y`).

``` r

output <- format_output(r, var_select = "E")
head(output)
#> # A tibble: 6 x 4
#>   replicate compartment     t     y
#>       <dbl> <chr>       <dbl> <dbl>
#> 1         1 E             0.1    20
#> 2         1 E             0.2    20
#> 3         1 E             0.3    20
#> 4         1 E             0.4    20
#> 5         1 E             0.5    20
#> 6         1 E             0.6    20
```

If we wanted age-disaggregated data, we could set `reduce_age` to
`FALSE` which will generate the same dataframe as before, but with an
additional column indicating the age-group.

``` r

output <- format_output(r, var_select = "E", reduce_age = FALSE)
head(output)
#> # A tibble: 6 x 5
#>   replicate age_group compartment     t     y
#>       <dbl>     <int> <chr>       <dbl> <dbl>
#> 1         1         1 E             0.1     0
#> 2         1         1 E             0.2     0
#> 3         1         1 E             0.3     0
#> 4         1         1 E             0.4     0
#> 5         1         1 E             0.5     0
#> 6         1         1 E             0.6     0
```

### 2\. Changing parameters in the model.

The model has a number of parameters for setting the R0, demography,
contact matrices, the durations of each compartment and the health care
outcomes and healthcare availability. In addition, the initial state of
the population can be changed as well as simulation parameters, such as
the number of replicates, length of simulation and the timestep. For a
full list of model inputs, please see the function
[documentation](https://mrc-ide.github.io/squire/reference/run_explicit_SEEIR_model.html)

For example, changing the initial R0 (default = 3), number of replicates
( default = 10), simulation length (default = 365 days) and time step
(default = 0.5 days), as well as setting the population and contact
matrix manually:

``` r

# Get the population
pop <- get_population("United Kingdom")
population <- pop$n

# Get the mixing matrix
contact_matrix <- get_mixing_matrix("United Kingdom")

# run the model
r <- run_explicit_SEEIR_model(population = population, 
                              contact_matrix_set = contact_matrix,
                              R0 = 2.5, 
                              time_period = 200,
                              dt = 0.1,
                              replicates = 5)
plot(r)
#> Warning in plot.squire_simulation(r): Summary statistic estimated from <10
#> replicates
#> Warning in plot.squire_simulation(r): Confidence bounds estimated from <10
#> replicates
```

<img src="man/figures/README-set params-1.png" width="100%" />

We can also change the R0 and contact matrix at set time points, to
reflect changing behaviour resulting from interventions. For example to
set an 80% reduction in the contact matrix after 100 days :

``` r

# run the model
r <- run_explicit_SEEIR_model(population = population, 
                              tt_contact_matrix = c(0, 100),
                              contact_matrix_set = list(contact_matrix,
                                                        contact_matrix*0.2),
                              R0 = 2.5, 
                              time_period = 200,
                              dt = 0.1,
                              replicates = 5)
plot(r, var_select = "infections")
#> Warning in plot.squire_simulation(r, var_select = "infections"): Summary
#> statistic estimated from <10 replicates
#> Warning in plot.squire_simulation(r, var_select = "infections"): Confidence
#> bounds estimated from <10 replicates
```

<img src="man/figures/README-set contact matrix decrease-1.png" width="100%" />

where `n_E2_I` is the daily number of new infections.

To show an 80% reduction after 80 days but only maintained for 40 days :

``` r

# run the model
r <- run_explicit_SEEIR_model(population = population, 
                              tt_contact_matrix = c(0, 80, 120),
                              contact_matrix_set = list(contact_matrix,
                                                        contact_matrix*0.2,
                                                        contact_matrix),
                              R0 = 2.5, 
                              time_period = 220,
                              dt = 0.1,
                              replicates = 5)
plot(r, var_select = "infections")
#> Warning in plot.squire_simulation(r, var_select = "infections"): Summary
#> statistic estimated from <10 replicates
#> Warning in plot.squire_simulation(r, var_select = "infections"): Confidence
#> bounds estimated from <10 replicates
```

<img src="man/figures/README-set contact matrix decrease and relax-1.png" width="100%" />

Alternatively, we could set a changing R0, which falls below 1 after 80
days:

``` r

# run the model
r <- run_explicit_SEEIR_model(population = population, 
                              contact_matrix_set = contact_matrix,
                              tt_R0 = c(0, 80),
                              R0 = c(2.5, 0.9),
                              time_period = 200,
                              dt = 0.1,
                              replicates = 5)
plot(r, var_select = "infections")
#> Warning in plot.squire_simulation(r, var_select = "infections"): Summary
#> statistic estimated from <10 replicates
#> Warning in plot.squire_simulation(r, var_select = "infections"): Confidence
#> bounds estimated from <10 replicates
```

<img src="man/figures/README-set R0 decrease-1.png" width="100%" />

The model in squire also allows you to alter healthcare capacity.
Default values for the arguments `hosp_bed_capacity` and `ICU_capacity`
are taken from the World Bank and a systematic review of the literature.
However, you can also specify your own:

``` r

library(patchwork)

r <- run_explicit_SEEIR_model(population = population, 
                              contact_matrix_set = contact_matrix,
                              R0 = 2.5, 
                              time_period = 200,
                              dt = 0.1,
                              replicates = 15,
                              hosp_bed_capacity = 1000,
                              ICU_bed_capacity = 100)

c <- plot(r, var_select = "hospital_occupancy")
d <- plot(r, var_select = "ICU_occupancy")
c / d 
```

<img src="man/figures/README-change healthcare capacity-1.png" width="100%" />

### 3\. Extracting and Plotting Relevant Outputs

Whilst the above provides quick and easy ways to plot relevant outputs,
we recognise users might want to play around with the data themselves.
This can be done with the `format_output` function.

``` r
library(ggplot2)
library(patchwork)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

x <- run_explicit_SEEIR_model(country = "Afghanistan", hosp_bed_capacity = 500, ICU_bed_capacity = 200)

deaths <- format_output(x = x, var_select = "deaths") %>%
  mutate(replicate = factor(replicate))
a <- ggplot(deaths, aes(x = t, y = y, col = replicate)) +
  geom_line() + ylab("Daily Deaths")

infections <- format_output(x = x, var_select = "infections") %>%
  mutate(replicate = factor(replicate))
b <- ggplot(infections, aes(x = t, y = y, col = replicate)) +
  geom_line() + ylab("Daily Infections")

hosp_bed <- format_output(x = x, var_select = "hospital_occupancy") %>%
  mutate(replicate = factor(replicate))
c <- ggplot(hosp_bed, aes(x = t, y = y, col = replicate)) +
  geom_line() + ylab("Hospital Bed Occupancy")

ICU_bed <- format_output(x = x, var_select = "ICU_occupancy") %>%
  mutate(replicate = factor(replicate))
d <- ggplot(ICU_bed, aes(x = t, y = y, col = replicate)) +
  geom_line() + ylab("ICU Bed Occupancy")

z <- a + b + c + d +
  plot_layout(guides = 'collect')
z
```

<img src="man/figures/README-untransformed-1.png" width="100%" />

Hospital bed occupancy fluctuates very slightly above the user-specified
number of hospital beds available (+1 in the above example). This is due
to individuals being discharged from ICU beds into general hospital beds
when general hospital bed occupancy is already at maximum capacity.

We assume that these individuals newly discharged from the ICU would in
reality be prioritised to receive general hospital beds and so give them
a general hospital bed. Whilst this is now shown in the above panel of
hospital bed occupancy, this means that less-sick individuals in general
hospital beds would be discharged slightly early to make space. We
assume that for these small number of the healthiest individuals in
hospital, there is no excess mortality associated with this early
discharge.

Future work will explore a range of bed-prioritisation structures to
consider how results might be shaped by who gets a bed and when.

### 4\. Calibrating the Model to Observed Deaths Data

The model can be simply calibrated to time series of deaths reported in
settings. This can be done using the `calibrate` function. For example,
let’s use the time series of deaths in Algeria up to the 22nd April
2020:

``` r
df <- read.csv(squire:::squire_file("extdata/example_DZA.csv"), stringsAsFactors = FALSE)
head(df)
#>   X                 date deaths cases
#> 1 1 2020-02-26T00:00:00Z      0     1
#> 2 2 2020-03-01T00:00:00Z      0     2
#> 3 3 2020-03-04T00:00:00Z      0     2
#> 4 4 2020-03-05T00:00:00Z      0     7
#> 5 5 2020-03-06T00:00:00Z      0     5
#> 6 6 2020-03-09T00:00:00Z      0     3
```

We can calibrate the model to this data. This calibration will infer the
most likely start date of the epidemic and the R0 at the start of the
epidemic. Calibration occurs by scanning across a range of possible R0
values and start dates. Fitting works using a particle filter, which can
be parallelised using `future::plan(future::multiprocess())` before
running `calibrate`.

``` r
# set up for parallelisation
future::plan(future::multiprocess())

# Fit model
out <- calibrate(
      data = df,
      R0_min = 2,
      R0_max = 4,
      R0_step = 0.5,
      first_start_date = "2020-02-10",
      last_start_date = "2020-02-22",
      day_step = 4,
      replicates = 10,
      n_particles = 20,
      country = "Algeria"
    )
#>  Progress: ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────              100% Progress: ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────              100% Progress: ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────        100% Progress: ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────        100% Progress: ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── 100% Progress: ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── 100%
#> 
#>  Progress: ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────              100% Progress: ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────              100% Progress: ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────              100% Progress: ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────              100% Progress: ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── 100%
```

`calibrate` returns the same output as `run_explicit_SEEIR_model`, with
the first three elements in `out` being the simulation outputs, the
model and model parameters. Note that simulation replicates are aligned
to the maximum date in the data provided, and as a result we can use the
same plotting functions as before:

``` r
plot(out, "deaths", date_0 = max(df$date), x_var = "date")
```

<img src="man/figures/README-plot particle deaths-1.png" width="100%" />

With default parameters, `calibrate` will simulate up to the maximum date
in the data provided. The fit to this data can be shown using the
plotting function and specifying `particle_fit` to be `TRUE`

``` r
plot(out, particle_fit = TRUE)
```

<img src="man/figures/README-particle fit no ints-1.png" width="100%" />

We can see that this is not the best fit, capturing only the beginning
of the data points. We can also diagnose the fitting further by looking
at the `scan_results` of the grid search.

``` r
plot(out$scan_results)
```

<img src="man/figures/README-show grid fits-1.png" width="100%" />

This grid shows the negative log likeihood for each parameter pair. We
can also plot the related probability for each pair as follows:

``` r
plot(out$scan_results, what = "probability")
```

<img src="man/figures/README-show grid fits probs-1.png" width="100%" />

The reason for the poor fits to the data shown earlier is because
Algeria has implemented interventions prior to today. These can also be
incorporated into `calibrate`. For example, we can grab the assumed
changes to transmission based on government intervention for
Algeria.

``` r
interventions <- read.csv(squire:::squire_file("extdata/example_DZA_intervention.csv"))
int_unique <- squire:::interventions_unique(interventions)
int_unique
#> $dates_change
#> [1] "2020-03-12" "2020-03-18" "2020-03-22" "2020-03-23"
#> 
#> $change
#> [1] 0.900 0.850 0.550 0.225
```

We can then provide these to `calibrate` as the dates (`date_R0_change`)
and relative reductions to R0 (`R0_change`). We will also specify for
model fits to be continued for 14 days into the future with `forceast
= 14`:

``` r
out <- calibrate(
      data = df,
      R0_min = 2,
      R0_max = 4,
      R0_step = 0.5,
      first_start_date = "2020-02-10",
      last_start_date = "2020-02-22",
      day_step = 4,
      replicates = 10,
      n_particles = 20,
      forecast = 14,
      R0_change = int_unique$change,
      date_R0_change = int_unique$dates_change,
      country = "Algeria"
    )
#>  Progress: ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────              100% Progress: ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── 100%
#> 

NEWS.md
# squire 0.7.0

* The `drjacoby` package can now be used to fit mcmc for deterministic model
* `drjacoby_mcmc` function analogous to `pmcmc` for fitting models
* `convert_log_likelihood_func_for_drjacoby` helps convert log likelihood
functions, e.g. `calc_loglikelihood` into suitable likelihood functions that can
be correctly passed as `log_likelihood` to `drjacoby_mcmc`.
* `convert_log_prior_func_for_drjacoby` does the same for priors
* The above changes are because `drjacoby` requires initial parameters and ranges
to be passed as a `data.frame` internally, which means dates are internally handled
as numerics, e.g. "2020-01-01" is 18262. Consequently, the sampled parameters from
`drjacoby_mcmc` will have `start_date` values as numerics. See `sample_drjacoby`
function for how these are handled correctly to align with converted log likelihood
function draws. 

# squire 0.6.11

* `parse_durations` bug fix. Handles dur_E, dur_IMild and dur_ICase correctly now

# squire 0.6.10

* `sample_pmcmc` bug fix dor negative posterior checks incorrectly. 

# squire 0.6.9

* `atol` and `rtol` can be passed to `pmcmc` fitting as named arguments in 
`pars_obs` to be used in `run_deterministic_comparison`
* `pmcmc` can now be used to fit to PCR prevalence data (deterministic model only) 
by passing `pcr_df` and `pcr_det` as named list elements of `pars_obs`

# squire 0.6.8

* `run_deterministic_comparison` patches for sero fitting and argument checks

# squire 0.6.7

* `pmcmc` can now be used to fit to serology data (deterministic model only) 
by passing `sero_df` and `sero_det` as names list elements of `pars_obs`

# squire 0.6.6

* `projections` can be used now for `nimue` models

# squire 0.6.5

* `pmcmc` now accepts dates of change for vaccine efficacy for `nimue` 

# squire 0.6.4

* `pmcmc` changes to adapt for `nimue` class models and vaccines

# squire 0.6.3

* `beta_est` accepts `nimue` models

# squire 0.6.2

* Typo in `dur_not_get_mv_survive`

# squire 0.6.1

* `pmcmc(..., scaling_factor)` allows scaling factors to be passed to pmcmc
* Correction to `deterministic_model` model class

# squire 0.6.0

* Waning added to squire

# squire 0.5.8

* Updated to use odin 1.1.0
* `odin_is_discrete` new unexported function

# squire 0.5.7

* Github actions included for automatic release tagging

# squire 0.5.6

* `pmcmc(..., gibbs_sampling = TRUE)` implements Gibbs sampler

# squire 0.5.5

* `population$n` are all integer class

# squire 0.5.4

* `beta_est` accepts apothecary models
* Updated json odin files that weren't compiled correctly

# squire 0.5.3

* `default_durations` contains all durations
* `default_durations` and `default_probs` now exported for ease
* `run_deterministic_SEIR_model` has odin pars exported now

# squire 0.5.2

* `Parameters` vignette updated to include new parameters
* `run_deterministic_comparison` bug fix for when transmission is eliminated 

# squire 0.5.1

* `projections` bug fixed for setting time varying durations
* Taiwan healthcare capacity added 
* `default_durations` updated for incorrect hospital durations

# squire 0.5.0

## 0.5.0 Major

* Parameter Updates. See parameters vignette for update.
* `triggering` function for doing `projections` with model state triggered R0 changes 
* Time-varying durations of ICU and hospital stay

## 0.5.0 Minor

* `projections` allows for different args to be passed to each replicate
* `projections` allows for additional user model args to be passed


# squire 0.4.34

* `pmcmc` argument `treated_deaths_only` will use deaths from treatment in likelihood

# squire 0.4.33

* `format_output` accepts `deaths_treatment` and `deaths_no_treatment`

# squire 0.4.32

* `intervention_dates_for_odin` fills dates earlier than `date_R0_change`

# squire 0.4.31

* Correction to random walk parameters to more accurately continue the walk

# squire 0.4.30

* `evaluate_Rt_pmcmc` takes `Rt_rw_n` for random walk esque shifts

# squire 0.4.29

* `format_output` backwards compatible for incidence

# squire 0.4.28

* `calibrate` patched as `dt` argument was not being passed through
* `format_output` corrected for hospital/ICU incidence

# squire 0.4.27 

* `evaluate_Rt_pmcmc` patch for non specified `Rt_shift` in 3 parameter version

# squire 0.4.26 

* `pmcmc` more flexible for varying which parameters are fitted.
* `evaluate_Rt_pmcmc` altered for increased flexibility

# squire 0.4.25

* `format_output` now returns `hospital_incidence` and `ICU_incidence`
* `evaluate_Rt_pmcmc` patch for `Meff_switch_date` earlier than `date_R0_change[1]`

# squire 0.4.24

* `Meff_pl` can be scaled or not in relation to Meff

# squire 0.4.23

* `pmcmc` no longer takes Rt_func as hard coded given Meff_pl complexity

# squire 0.4.22

* Johnstone Change approach implemented for pmcmc RM

# squire 0.4.21

* `evaluate_Rt` patch for `roll` of Meff

# squire 0.4.20

* `evaluate_Rt` allows for `date_Meff_change` to be later/earlier than `date_R0_change` ends

# squire 0.4.19

* Large overhaul to make `projections` and plotting work with deterministic models

# squire 0.4.18

* `calc_loglikelihood` fix for reduced date vectors

# squire 0.4.17

* `pars_init` for `pmcmc` can be provided as a list for where each chain starts

# squire 0.4.16

* `pmcmc` method for 4 parameter estimation.

# squire 0.4.15

* `projections` now matches by column names when including projection outputs
* Patch to `t0_variables` for new Rt_func approach

# squire 0.4.14

* replace chained binomials with mutivariate hypergeometric sampling in stochastic model
* stochastic model to output number newly requiring ICU beds at each timepoint

# squire 0.4.13

* `calibrate` now takes an `Rt_func` argument for specifying Rt calculation

# squire 0.4.12

* Explicit Meff inclusion in `R0_date_particle_filter` and `projections` fix

# squire 0.4.11

* patching `projections` when projecting further than original calibrate run

# squire 0.4.10

* `calibrate` new argument for `R0_prior`

# squire 0.4.9


tester.R
# Load Required Libraries
library(patchwork); library(dplyr); library(ggplot2)

# Run an Instance of the Model
set.seed(2)
x <- run_explicit_SEEIR_model(country = "Afghanistan", hosp_bed_capacity = 500, ICU_bed_capacity = 600)
index <- squire:::odin_index(x$model)

# Checking Hosp Bed Capacity
check_hosp_capacity <- x$output[, index$hosp_bed_capacity, 1]

# Checking Oxygen Assignments
number_get_ox <- apply(x$output[, index$number_get_Ox, 1], 1, sum)
number_get_ox_live <- apply(x$output[, index$n_IOxGetLive1, 1], 1, sum)
number_get_ox_die <- apply(x$output[, index$n_IOxGetDie1, 1], 1, sum)
number_get_ox_comb <- number_get_ox_live + number_get_ox_die
which(number_get_ox_comb != number_get_ox)
number_not_get_ox <- apply(x$output[, index$number_notget_Ox, 1], 1, sum)
number_requiring_ox <- apply(x$output[, index$number_requiring_Ox, 1], 1, sum)
number_get_ox <- apply(x$output[, index$number_get_Ox, 1], 1, sum)
number_ox <- number_not_get_ox + number_get_Ox
which(number_ox != number_requiring_ox)

# Checking That All Newly Hospitalised Individuals Are Correctly Assigned to ICU or Hosp Beds
newly_hospitalised <- apply(x$output[, index$n_ICase2_Hosp, 1], 1, sum)
total_number_requiring_ICU <- x$output[, index$total_number_requiring_IMV, 1]
total_number_requiring_ox <- x$output[, index$total_number_requiring_Ox, 1]
tot <- total_number_requiring_ICU + total_number_requiring_ox
which(newly_hospitalised != tot)

# Spotting for Bugs in Hospital Occupancy
hosp_occ <- x$output[, index$hosp_occ, 1]
total_number_requiring_ox <- x$output[, index$total_number_requiring_Ox, 1]
current_free_hosp <- x$output[, index$current_free_hosp, 1]
total_number_get_hosp <- x$output[, index$total_number_get_hosp, 1]
number_get_Ox <- apply(x$output[, index$number_get_Ox, 1], 1, sum)

max(hosp_occ)
min(current_free_hosp)
max_index <- which(hosp_occ == max(hosp_occ))
start <- max_index - 10
end <- max_index + 9

bloop <- x$output[, index$temp5, 1]
plot(bloop, type = "l")

which(bloop < 0)
which(hosp_occ > 500)

hosp_occ[start:end]
current_free_hosp[start:end]

cbind(hosp_occ[start:end], current_free_hosp[start:end], total_number_get_hosp[start:end])
total_number_get_hosp[start:end]
total_number_requiring_ox[start:end]

ioxgetlive1 <- apply(x$output[, index$IOxGetLive1, 1], 1, sum)
ioxgetlive2 <- apply(x$output[, index$IOxGetLive2, 1], 1, sum)
ioxgetdie1 <- apply(x$output[, index$IOxGetDie1, 1], 1, sum)
ioxgetdie2 <- apply(x$output[, index$IOxGetDie2, 1], 1, sum)
irec1 <- apply(x$output[, index$IRec1, 1], 1, sum)
irec2 <- apply(x$output[, index$IRec2, 1], 1, sum)
overall <- ioxgetlive1 + ioxgetlive2 + ioxgetdie1 + ioxgetdie2 + irec1 + irec2
max(overall)
plot(overall, type = "l")






# ioxgetlive1[(max_index-2):max_index]
# ioxgetlive2[(max_index-2):max_index]
# ioxgetdie1[(max_index-2):max_index]
# ioxgetdie2[(max_index-2):max_index]
# irec1[(max_index-2):max_index]
# irec2[(max_index-2):max_index]
# overall[(max_index-2):max_index]
#
# plot(ioxgetlive1, type = "l")
# plot(ioxgetlive2, type = "l")
# plot(ioxgetdie1, type = "l")
# plot(ioxgetdie2, type = "l")
# plot(irec1, type = "l")
# plot(irec2, type = "l")
#
#

#### scrap for now ####

# which(total_number_get_hosp != number_get_Ox)
# plot(hosp_occ, type = "l")
# max(hosp_occ)
# plot(current_free_hosp, type = "l")
# min(current_free_hosp)
# plot(total_number_get_hosp, type = "l")
#
# x <- c()
# for (i in 1:length(current_free_hosp)) {
#   x[i] <- if (current_free_hosp[i] <= 0) 0 else (if(current_free_hosp[i] - total_number_requiring_ox[i] >= 0) total_number_requiring_ox[i] else(current_free_hosp[i]))
# }
# table(x - total_number_get_hosp)
#
# min(total_number_get_hosp)
# which(number_get_Ox != total_number_requiring_ox)
# a <- current_free_hosp[which(number_get_Ox != total_number_requiring_ox)]
# b <- total_number_requiring_ox[which(number_get_Ox != total_number_requiring_ox)]
# which(a > b)
#
# plot(total_number_requiring_ox)
#
# total_number_get_hosp[total_number_requiring_ox > 0]
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
# #
# ICU_occ <- x$output[, index$ICU_occ, 1]
# total_number_requiring_ICU <- x$output[, index$total_number_requiring_IMV, 1]
# current_free_ICU <- x$output[, index$current_free_ICUs, 1]
# total_number_get_IMV <- x$output[, index$total_number_get_IMV, 1]
# number_get_ICU <- apply(x$output[, index$number_get_IMV, 1], 1, sum)
#
# which(total_number_get_IMV != number_get_ICU)
# plot(ICU_occ, type = "l")
# max(ICU_occ)
# plot(current_free_ICU, type = "l")
# min(current_free_ICU)
# plot(total_number_get_IMV, type = "l")
# which(number_get_ICU != total_number_requiring_ICU)
# a <- current_free_ICU[which(number_get_ICU != total_number_requiring_ICU)]
# b <- total_number_requiring_ICU[which(number_get_ICU != total_number_requiring_ICU)]
# which(a > b)
#
#
#
#
#
#
#
# plot(hosp_occ, type = "l")
# max(hosp_occ)
# plot(current_free_hosp, type = "l")
# min(current_free_hosp)
#
# total_number_get_hosp[which(current_free_hosp <= 0)]
# total_number_get_hosp[which(current_free_hosp - total_number_requiring_ox >= 0)]
#
# total_number_get_hosp[which(!(current_free_hosp <= 0) | !(current_free_hosp - total_number_requiring_ox >= 0))]
#
# total_number_get_hosp[950:1000] - current_free_hosp[950:1000]
#
# total_number_get_hosp[which(hosp_occ >= 500)]
# current_free_hosp[which(hosp_occ >=500)]
#
#
# current_free_hosp
# min(current_free_hosp)
#
#
# plot(free, type = "l")
#
# number_get_hosp[which(free < 0)]
# number_get_hosp[which(free - req_ox >= 0)]
# number_get_hosp[which(free - req_ox < 0)]
#
#
#
# unique(number_get_hosp[which(free - req_ox > 0)])
#
# plot(free, type = "l")
# plot(number_get_hosp, type = "l")
#
# plot(free + number_get_hosp, type = "l")
#
# max(free + number_get_hosp)
# min(free + number_get_hosp)
#
# max(free)
# min(free)
#
# free[free < 0] <- 0
# a <- free - get_ox
#
#
# plot(hosp_occ, type = "l")
# max(hosp_occ)
# which(hosp_occ == 514)
#
# plot(hosp_occ, type = "l")
# plot(a, type = "l")
# unique(hosp_occ + a)
#
# total_get_ox <- x$output[, index$total_number_get_hosp, 1]
# get_ox <- apply(x$output[, index$number_get_Ox, 1], 1, sum)
# unique(total_get_ox - get_ox)
#
# total_get_ox[715]
# get_ox[715]
# free[715]
#
# rec <- x$output
#
#
# unique(a)
# plot(a, type = "l")
# min(a)
#
# require_ox <- apply(x$output[, index$number_requiring_Ox, 1], 1, sum)
#
#
# rec <- max(apply(x$output[, c(index$IRec1, index$IRec2), 1], 1, sum))
#
# hosp_bed <- format_output(x = x, var_select = "hospital_occupancy") %>%
#   mutate(replicate = factor(replicate)) %>%
#   filter(replicate == 1)
# max(hosp_bed$y)
#
# colnames(x$output[, , 1])
#
# deaths <- format_output(x = x, var_select = "deaths") %>%
#   mutate(replicate = factor(replicate))
# a <- ggplot(deaths, aes(x = t, y = y, col = replicate)) +
#   geom_line() + ylab("Daily Deaths")
#
# infections <- format_output(x = x, var_select = "infections") %>%
#   mutate(replicate = factor(replicate))
# b <- ggplot(infections, aes(x = t, y = y, col = replicate)) +
#   geom_line() + ylab("Daily Infections")
#
# hosp_bed <- format_output(x = x, var_select = "hospital_occupancy") %>%
#   mutate(replicate = factor(replicate))
# c <- ggplot(hosp_bed, aes(x = t, y = y, col = replicate)) +
#   geom_line() + ylab("Hospital Bed Occupancy")
#
# ICU_bed <- format_output(x = x, var_select = "ICU_occupancy") %>%
#   mutate(replicate = factor(replicate))
# d <- ggplot(ICU_bed, aes(x = t, y = y, col = replicate)) +
#   geom_line() + ylab("ICU Bed Occupancy")
#
# z <- a + b + c + d +
#   plot_layout(guides = 'collect')
# z
